src/lib.rs

```diff
@@ -8323,6 +8421,74 @@ mod tests {
         );
     }
 
+    #[test]
+    /// Tests that shutting down a stream restores flow control for unsent data.
+    fn stream_shutdown_write_unsent_tx_cap() {
+        let mut config = Config::new(crate::PROTOCOL_VERSION).unwrap();
+        config
+            .load_cert_chain_from_pem_file("examples/cert.crt")
+            .unwrap();
+        config
+            .load_priv_key_from_pem_file("examples/cert.key")
+            .unwrap();
+        config
+            .set_application_protos(b"\x06proto1\x06proto2")
+            .unwrap();
+        config.set_initial_max_data(15);
+        config.set_initial_max_stream_data_bidi_local(30);
+        config.set_initial_max_stream_data_bidi_remote(30);
+        config.set_initial_max_stream_data_uni(30);
+        config.set_initial_max_streams_bidi(3);
+        config.set_initial_max_streams_uni(0);
+        config.verify_peer(false);
+
+        let mut pipe = testing::Pipe::with_config(&mut config).unwrap();
+        assert_eq!(pipe.handshake(), Ok(()));
+
+        // Client sends some data.
+        assert_eq!(pipe.client.stream_send(4, b"hello", true), Ok(5));
+        assert_eq!(pipe.advance(), Ok(()));
+
+        let mut r = pipe.server.readable();
+        assert_eq!(r.next(), Some(4));
+        assert_eq!(r.next(), None);
+
+        let mut b = [0; 15];
+        assert_eq!(pipe.server.stream_recv(4, &mut b), Ok((5, true)));
+
+        // Server sends some data.
+        assert_eq!(pipe.server.stream_send(4, b"hello", false), Ok(5));
+        assert_eq!(pipe.advance(), Ok(()));
+
+        // Server buffers some data, until send capacity limit reached.
+        assert_eq!(pipe.server.stream_send(4, b"hello", false), Ok(5));
+        assert_eq!(pipe.server.stream_send(4, b"hello", false), Ok(5));
+        assert_eq!(
+            pipe.server.stream_send(4, b"hello", false),
+            Err(Error::Done)
+        );
+        assert_eq!(
+            pipe.server.stream_send(8, b"hello", false),
+            Err(Error::Done)
+        );
+
+        // Server shuts down stream.
+        assert_eq!(pipe.server.stream_shutdown(4, Shutdown::Write, 42), Ok(()));
+        assert_eq!(pipe.advance(), Ok(()));
```

Comment:
Does this unintentionally cause the client to ack or MAX_DATA?

Reply:
In other words, are we sure the following send is allowed because we reset, and not because the client just granted more capacity

Reply:
`stream_shutdown()` drops buffers before anything is sent, and previously we only sent 5 bytes (less than half of the fc window), so there shouldn't be any MAX_DATA. But FWIW, while making the tests I was running with the `tx frm ...` logs enabled and I didn't see any MAX_DATA (the server does seem to send DATA_BLOCKED anyway though, but that should be harmless I think, and our code just ignores it anyway).

Reply:
I added `ssert_eq!(pipe.client.should_update_max_data(), false);` just to make sure. The value shouldn't change during `advance()` since there is no data being sent.